posts/2023-10-31-23.0.0.11-beta.adoc

```diff
@@ -0,0 +1,330 @@
+
+MicroProfile Reactive Messaging and Reactive Stream Operators 3.0 provide a range of new features and is compatible with Jakarta EE 9 onwards.
+
+To use both features together, add the Microprfile Reactive Messaging 3.0 feature to the `server.xml`:
+
+[source,xml]
+----
+<feature>mpReactiveMessaging-3.0</feature>
+----
+
+Enabling MicroProfile Reactive Messaging 3.0 automatically enables the MicroProfile Reactive Stream Operators 3.0 feature too.
+
+To use MicroProfile Reactive Stream Operators 3.0 without MicroProfile Reactive Messaging:
+
+[source,xml]
+----
+<feature>mpReactiveStreams-3.0</feature>
+----
+
+
+=== Negative acknowledgements
+
+In MicroProfile Reactive Messaging 1.0, Messages could only be positively acknowledged. If there was a problem with the payload or if exceptional behavior occurred, there was no mechanism to indicate the problem or to handle the problem if it occurred within the stream. The new _nack_ support can send or handle these events.
+
+To explicitly negatively acknowledge an incoming Message:
+
+[source,java]
+----
+@Incoming("data")
+@Outgoing("out")
+public Message<String> process(Message<String> m) { 
+    String s = m.getPayload();
+    if (s.equalsIgnoreCase("b")) {
+        // we cannot fail, we must nack explicitly.
+        m.nack(new IllegalArgumentException("b"));
+        return null;
+    }
+    return m.withPayload(s.toUpperCase());
+}
+----
+
+The method signature of accepting a Message indicates that the acknowledgement strategy for this method is `MANUAL`. It is the responsibility of your code to `ack()` or `nack()` the message. In the above example, the message can be acknowledged downstream of the out channel.
+
+To throw an exception that causes a negative acknowledgement:
+
+[source,java]
+----
+@Incoming("data")
+@Outgoing("out")
+public String process(String s) {  
+    if (s.equalsIgnoreCase("b")) {
+        throw new IllegalArgumentException("b"); 
+    }
+    return s.toUpperCase();
+}
+----
+
+The method signature of accepting a payload indicates that the acknowledgement strategy for this method is `POST_PROCESSING`. The implementation handles `ack()` and `nack()` calls on the message after the method or chain completes. The upstream data receives the negative acknowledgement with the reason of `IllegalArgumentException`.
+
+=== Emitters
+
+One significant limitation of MicroProfile Reactive Messaging 1.0 was how to integrate imperative code such as RESTful resources and beans using Reactive Messaging which would put and take messages on to a channel according to the `Outgoing` or `Incoming` annotations. In version 3.0, emitters provide a bridge across the two models.
```

Comment:
> One significant limitation of MicroProfile Reactive Messaging 1.0 was how to integrate imperative code such as RESTful resources and beans using Reactive Messaging which would put and take messages on to a channel according to the `Outgoing` or `Incoming` annotations.

this sentence seems confusing- maybe something like...

```suggestion
MicroProfile Reactive Messaging 1.0 did not offer a clear way to integrate imperative code, such as RESTful resources and beans, because Reactive Messaging put and took messages onto a channel according to the `Outgoing` or `Incoming` annotations. In version 3.0, emitters provide a bridge across the two models.
```